app/src/main/java/com/example/studentarena/EndlessRecyclerViewScrollListener.java

```diff
@@ -0,0 +1,88 @@
+package com.example.studentarena;
+
+import androidx.recyclerview.widget.GridLayoutManager;
+import androidx.recyclerview.widget.LinearLayoutManager;
+import androidx.recyclerview.widget.RecyclerView;
+import androidx.recyclerview.widget.StaggeredGridLayoutManager;
+
+public abstract class EndlessRecyclerViewScrollListener extends RecyclerView.OnScrollListener{
+    // The minimum amount of items to have below your current scroll position
+    // before loading more.
+    private int visibleThreshold = 5;
+    // The current offset index of data you have loaded
+    private int currentPage = 0;
+    // The total number of items in the dataset after the last load
+    private int previousTotalItemCount = 0;
+    // True if we are still waiting for the last set of data to load.
+    private boolean loading = true;
+    // Sets the starting page index
+    private int startingPageIndex = 0;
+
+    RecyclerView.LayoutManager mLayoutManager;
+
+    public EndlessRecyclerViewScrollListener(LinearLayoutManager layoutManager) {
+        this.mLayoutManager = layoutManager;
+    }
+
+    public EndlessRecyclerViewScrollListener(GridLayoutManager layoutManager) {
+        this.mLayoutManager = layoutManager;
+        visibleThreshold = visibleThreshold * layoutManager.getSpanCount();
+    }
+
+    public EndlessRecyclerViewScrollListener(StaggeredGridLayoutManager layoutManager) {
+        this.mLayoutManager = layoutManager;
+        visibleThreshold = visibleThreshold * layoutManager.getSpanCount();
+    }
```

Comment:
do you need both of these?

i think this is an example of Object-Oriented Programming. `StaggeredGridLayoutManager` is a _subclass_ of the more general `GridLayoutManager` _parent class_ (a.k.a _superclass_). if a function just says it takes a `GridLayoutManager` argument, then any subclass or implementation of `GridLayoutManager` will work fine.

the classic example is something like this:
```
class Animal {
    void speak();
}

class Cat {
    @Override
    void speak() {
        System.out.println("meow");
    }
}
class Dog {
    @Override
    void speak() {
        System.out.println("Woof");
    }
}

class Cow {
    @Override
    void speak() {
        System.out.println("moo");
    }
}

class Main {
    public static void main() {
        Animal animals[] = {new Dog(), new Cat(), new Cow()};
        for (int i = 0; i < animals.length; ++i) {
            animals[i].speak();
        }
    }
}
```

our array of `Animal`s can contain any type of `Animal`. and i think you can write a single `EndlessRecyclerViewScrollListener()` constructor that takes any type of `GridLayoutManager`.

Reply:
Yes I need the StaggeredGridLayoutManager layoutManager) as there will be no suitable constructor found for EndlessRecyclerViewScrollListener if I used the parent class GridLayoutManager.

Reply:
hmm... that doesn't make sense to me but i trust that you tried it so okay haha